Add files via upload

LukshaMaria opened the pull request:
Task 1-3, the code was split onto separete files.

LukshaMaria committed:
Add files via upload

Task 1-3, the code was split onto separete files.

LukshaMaria committed:
Delete usertable.js

LukshaMaria committed:
Add files via upload

HelgaListopad commented on modules/charts.js:
```diff
@@ -0,0 +1,17 @@
+const userchart = {
+    view: "chart",
+    url: "http://localhost/Projects-Webix/data/users.js",
```
Здесь и **везде**, где грузятся данные: нельзя писать абсолютные пути для файлов, которые вложены в тот же проект - нужно переделать на относительные внутри проекта. Сработает `"./data/users.js"`
У нас с тобой под localhost разная файловая система, и "localhost/Projects-Webix" для меня не существует

LukshaMaria replied:
fixed
Changed files: maintable.js, usertable.js, charts.js

LukshaMaria replied:
fixed

HelgaListopad commented on main.js:
```diff
@@ -0,0 +1,40 @@
+        ]
+    });
+    $$("list1").select("Dashboards");
+    $$("form1").bind("datatable1");
```
`bind` нужно убрать, так как в этом задании ты сама пишешь логику сохранения данных для тамблицы и формы

LukshaMaria replied:
fixed

LukshaMaria replied:
fixed

HelgaListopad commented on main.js:
```diff
@@ -0,0 +1,40 @@
+const label = {
+    view: "label",
+    label: "The software is provided by <a href='#'>https://webix.com</a>. All right reserved (c)",
+    align: "center"
+};
+const mainpart = {
+    cells: [
+        { id: "Dashboards", cols: [maintable, form] },
+        { id: "Users", cols: [usertable, userchart] },
+        { id: "Products", cols: [producttable] },
```
Мне не очень нравится, что в main.js используются переменные, которых в самом файле нет.
Либо нужно всё же перевести всё на es6 импорты (тогда ты в main будешь импортировать всё остальное и все переменные будут доступны в явном виде), либо код из main.js всё же перенести в основной html-файл

LukshaMaria replied:
fixed

LukshaMaria replied:
fixed

HelgaListopad replied:
Не совсем (ничего не работает) - оставила отдельный комментарий

HelgaListopad commented on main.js:
```diff
@@ -0,0 +1,40 @@
+    ]
+};
+webix.ready(function () {
+    const popup1 = webix.ui({
```
Если переменная `popup1` нигде не используется, можно не объявлять её и просто сразу инитить popup как `webix.ui({ ... })`

LukshaMaria replied:
fixed

LukshaMaria replied:
fixed

HelgaListopad commented on main.js:
```diff
@@ -0,0 +1,40 @@
+const mainpart = {
+    cells: [
+        { id: "Dashboards", cols: [maintable, form] },
+        { id: "Users", cols: [usertable, userchart] },
```
На иллюстрации `rows`

LukshaMaria replied:
В задании было написано 4 ячейки: Multiview should have 4 cells:...

HelgaListopad replied:
Речь о внешнем виде второй ячейки - комментарий оставлен только для неё
```js
{ id: "Users", rows: [
   userlist,    // here: toolbar with filter input/sorting buttons, list
   userchart
]},
```

LukshaMaria replied:
В задании используется cells

LukshaMaria replied:
аа, хорошо, исправлю

LukshaMaria replied:
fixed

HelgaListopad commented on modules/forms.js:
```diff
@@ -0,0 +1,87 @@
+        { view: "text", name: "rank", label: "Rank" },
+        { view: "text", name: "category", label: "Category" },
+        {
+            margin: 5, cols: [
```
Оставь только кнопки Save и Clear

LukshaMaria replied:
fixed

HelgaListopad commented on modules/forms.js:
```diff
@@ -0,0 +1,87 @@
+                    }
+                },
+                {
+                    view: "button", value: "Save", click: function () {
```
Выдели эту кнопку стилем webix_primary

LukshaMaria replied:
fixed

HelgaListopad commented on modules/forms.js:
```diff
@@ -0,0 +1,87 @@
+                        }
+                        else {
+                            list.add(item_data);
+                        }
```
Для того, чтобы начать это делать, должна сначала успешно пройти валидация полей

LukshaMaria replied:
fixed

HelgaListopad commented on modules/maintable.js:
```diff
@@ -0,0 +1,28 @@
+    ],
+    onClick: {
+        "wxi-trash": function (e, id) {
+            this.remove(id);
```
`remove` можно заментиь на confirm + remove из кнопки Delete (в задании нет, но есть смысл сохранить такое поведение)
```js
webix.confirm("Do you want to delete this item?", "confirm-warning")
.then(() => {
    this.remove(id);
});
```

LukshaMaria replied:
fixed

HelgaListopad replied:
`return false` нужно вынести из `webix.confirm` (т.е. вернуть на старое место)
Отдельно пометила, что там не закрыта скобка.

В целом в такой записи 
<details><summary>(кликни, чтобы развернуть)</summary>

```js
        "wxi-trash": function (e, id) {
            const list = $$("datatable1");
            webix.confirm("Do you want to delete this item?", "confirm-warning").then(function () {
                list.remove(id);
                return false;
            });
        }
```

</details>
обработчик клика ничего не возвращает (return написан для webix.confirm и никак не используется)
Более того, `confirm` асинхронный (ждёт действия от пользователя, потом вызывается `then`) и никак не прерывает дальнейшие действия, которые выполняются в обработчике клика или после него в коде webix

HelgaListopad commented on modules/maintable.js:
```diff
@@ -0,0 +1,28 @@
+    url: "http://localhost/Projects-Webix/data/data.js",
+    scroll: "y",
+    select: true,
+    autoConfig: true,
```
Стоит убрать - `autoConfig` нужен в первую очередь для случая, когда колонки не заданы в явном виде

LukshaMaria replied:
fixed

HelgaListopad commented on modules/maintable.js:
```diff
@@ -0,0 +1,28 @@
+    scroll: "y",
+    select: true,
+    autoConfig: true,
+    topSplit: 2,
```
Лучше убрать - в задании нет.

Кроме того, обрати внимание, что в колонке Id на закреплённые строки не срабатывает стиль "colls_css" - это тот случай, когда родной стиль (на ячейках появляется `webix_topcell`) перебивает тот, что написан для колонки. Исправляются такие вещи верным css-селектором.

LukshaMaria replied:
fixed.
Убрала autoConfig,topSplit и navigation и стиль применился ко всей колонке.

HelgaListopad commented on modules/usertable.js:
```diff
@@ -0,0 +1,43 @@
+    {
+        id: "usertable1",
+        view: "datatable",
+        url: "http://localhost/Projects-Webix/data/users.js",
```
"./data/users.js"

LukshaMaria replied:
fixed

HelgaListopad commented on modules/usertable.js:
```diff
@@ -0,0 +1,43 @@
+    },
+    {
+        id: "usertable1",
+        view: "datatable",
```
Здесь по заданию должен быть лист

LukshaMaria replied:
fixed

HelgaListopad commented on modules/usertable.js:
```diff
@@ -0,0 +1,43 @@
+    rows: 
+    [
+    {
+        cols: [
```
не хватает поля, через которое можно было бы фильтровать лист (`onTimedKeyPress` => `list.filter()`)

LukshaMaria replied:
fixed

HelgaListopad commented on modules/usertable.js:
```diff
@@ -0,0 +1,43 @@
+                return false;
+            }
+        },
+        scheme: {
+            $change: function (item) {
+                {
+                    if (item.id < 6) {
+                        item.$css = "highlight";
+                    }
+                }
+            }
+        }
```
Проверка id сработает здесь, но не очень хороша в целом - это может быть любое число или строка, поэтому лучше сразу учиться другим вариантам.

То же самое можно было бы сделать после загрузки данных в ready, работая с индексом строки, а не с id (id всё равно нужно, чтобы добавить стиль, но найти нужную строку лучше через индекс)
```js
ready() { 
   for (let i = 0; i < 5; i++) {
       const id = this.getIdByIndex(i);
      this.addCss(id, "highlight");       
   }
}

LukshaMaria replied:
fixed

HelgaListopad commented on modules/producttable.js:
```diff
@@ -0,0 +1,37 @@
+        },
+        { id: "price", header: "Price", width: 100 }
+    ],
+    data: [
```
загрузи из data/products.js

LukshaMaria replied:
fixed

HelgaListopad commented on modules/maintable.js:
```diff
@@ -0,0 +1,28 @@
+    id: "datatable1",
+    hover: "myhover",
+    view: "datatable",
+    url: "http://localhost/Projects-Webix/data/data.js",
```
./data/data.js

LukshaMaria replied:
fixed

LukshaMaria replied:
fixed

HelgaListopad commented:
Дополнительно (попробуй это только после того, как внесёшь все остальные изменения, в том числе переход на ES6 imports)

В разных файлах используются ID виджетов, которых там нет - например, popup c `id: "popup1"` объявлен в одном файле, а обращение к нему через это ID происходит в файле c тулбаром/кнопкой:
```js
// toolbar.js
click() {
     const node = this.getNode();
     $$("popup1").show(node);
}
```

Для любого проекта это не очень удобный и рискованный вариант, так как легко поменять ID в одном месте, упустив где-то ещё.

Чтобы избежать проблем, безопасным варирантом будет создать отдельный файл, откуда будут экспортироваться константы-айдишки, например
```js
export const list_id = "list";
export const popup_id = "popup";

// или единым объектом (примеры ниже для этого случая)
export const globalID = { 
   popup: "popup",
   list: "list"
}

```

Тогда их можно импортировать и переиспользовать во всех других файлах - везде будет одна и та же константа
```js
// main.js
import { globalID } from "/constants.js"
/* ... */
webix.ui({ 
   view: "popup",
   id: globalID.popup
})

```js
// toolbar.js
import { globalID} from "/constants.js"
/* ... */
click() {
     $$(globalID.popup).show();
}
```

HelgaListopad commented on task-webix.html:
```diff
@@ -1,133 +1,37 @@
+        }
+
+        .colls_css {
+            background-color: rgb(164, 164, 164);
```
Здесь лучше возьми цвет header'a таблицы

LukshaMaria replied:
fixed

LukshaMaria replied:
fixed

LukshaMaria committed:
Update charts.js

Changed file path.

LukshaMaria committed:
Update maintable.js

Changed file path

LukshaMaria committed:
Update usertable.js

Change file path

LukshaMaria committed:
Update main.js

Delete $$("form1").bind("datatable1");

LukshaMaria committed:
Update charts.js

Add function for export one value

LukshaMaria committed:
Update forms.js

Add function for export one value.

LukshaMaria committed:
Update maintable.js

Add function for export one value.

LukshaMaria committed:
Update menu.js

Add function for export one value.

LukshaMaria committed:
Update producttable.js

Add function for export one value.

LukshaMaria committed:
Update toolbar.js

Add function for export one value.

LukshaMaria committed:
Update usertable.js

Add function for export value

LukshaMaria committed:
Update task-webix.html

Delete files js-paths

LukshaMaria committed:
Update main.js

Add functiond for import files

LukshaMaria committed:
Update main.js

Delete value popup1

LukshaMaria committed:
Update main.js

LukshaMaria committed:
Update forms.js

Delete buttons "Add new", "Delete"

LukshaMaria committed:
Update forms.js

Add css style "webix_primary"

LukshaMaria committed:
Update maintable.js

Delete autoconfig

LukshaMaria committed:
Update style.css

Insert styles which were in file tasl-webix.html

LukshaMaria committed:
Update task-webix.html

Delete <style></style>

LukshaMaria committed:
Update forms.js

Add validation to button "Save"

LukshaMaria committed:
Update maintable.js

LukshaMaria committed:
Update maintable.js

Delete autoConfig,topSplit и navigation

LukshaMaria committed:
Update producttable.js

Added file path, deleted load information using "data".

LukshaMaria committed:
Update usertable.js

Changed logic

LukshaMaria committed:
Update maintable.js

Add form:"form1"

LukshaMaria committed:
Update main.js

Replaced cols to rows

LukshaMaria committed:
Update main.js

LukshaMaria committed:
Update maintable.js

Made connection datatable with form

HelgaListopad commented on task-webix.html:
```diff
@@ -1,133 +1,17 @@
-			$$("form1").bind("datatable1")
-		})
-	</script>
+    <script type="text/javascript" src="main.js"></script>
```
Код вообще не работает - здась скрипт нужно подключать как модуль (`type="module"`)

LukshaMaria replied:
Изменения внесены в код

HelgaListopad commented on modules/maintable.js:
```diff
@@ -0,0 +1,35 @@
+        { id: "del", template: "{common.trashIcon()}" }
+    ],
+    onClick: {
+        "wxi-trash": function (e, id) {
+            const list = $$("datatable1");
+            webix.confirm("Do you want to delete this item?", "confirm-warning").then(function () {
+                list.remove(id);
+                return false;
+       });
```
Не закрыта скобка => всё приложение падает с ошибкой

LukshaMaria replied:
Скобку } добавила

LukshaMaria replied:
Изменила на стрелочную функцию

HelgaListopad commented on modules/forms.js:
```diff
@@ -0,0 +1,69 @@
+                    view: "button", value: "Save", css:"webix_primary", click: function () {
+                        const list = $$("datatable1");
+                        const form = $$("form1");
+                        const item_data = form.getValues();
+                        if (item_data.id && form.validate()) {
+                            list.updateItem(item_data.id, item_data);
+                            webix.message("Information was changed to datatable!");
+                        }
+                        else if (form.validate()){
+                            list.add(item_data);
+                            webix.message("Information was added to datatable!");
+                        }
```
Стоит упростить
```js
if (form.validate()) {
   const item_data = form.getValues();
   if (item_data.id) { ... }
   else { ... }
}
```

LukshaMaria replied:
Изменения внесены в код.

HelgaListopad commented on modules/usertable.js:
```diff
@@ -0,0 +1,54 @@
+const usertable =
```
Это больше не таблица - есть смысл переименовать и переменную (включая импорт/экспорт), и сам файл

LukshaMaria replied:
Изменения внесены в код.

HelgaListopad commented on modules/usertable.js:
```diff
@@ -0,0 +1,54 @@
+                    function () {
+                        $$("usertable1").remove(id);
+                        return false;
+                        webix.message("Deleted");
```
Код после `return false` уже не выполняется

LukshaMaria replied:
Изменения внесены в код.

HelgaListopad commented on modules/usertable.js:
```diff
@@ -0,0 +1,54 @@
+                    function () {
+                        webix.message("Error");
+                    }
+                );
```
После webix.confirm не хватает return false, чтобы запретить селект по клику на эту иконку

LukshaMaria replied:
Изменения внесены в код.

HelgaListopad commented on modules/usertable.js:
```diff
@@ -0,0 +1,54 @@
+        id: "usertable1",
+        view: "list",
+        url: "./data/users.js",
+        template:"#name# from #country# <span class='webix_icon mdi mdi-close remove-icon' title='Remove'></span>",  
```
Я бы предложила выровнять иконки по правому краю
В стилях:
```css
.remove-icon {
    float: right;
    margin-top: 5px;
}
```

LukshaMaria replied:
Изменения внесены в код.

HelgaListopad commented on modules/maintable.js:
```diff
@@ -0,0 +1,35 @@
+                return false;
+       });
+    },
+    ready(){
+        $$("datatable1").attachEvent("onAfterSelect", function(id){
+            var item = this.getItem(id);
+            $$("form1").setValues(item);
+        })
+    }
```
Через ready вызывать attachEvent несколько излишне
В кофигурации таблицы лучше использовать `on`
```js
on: {
   onAfterSelect(id) { ... }
}
```

LukshaMaria replied:
Изменила ready на on

LukshaMaria committed:
modified:   main.js
modified:   modules/forms.js
modified:   modules/maintable.js
renamed:    modules/usertable.js -> modules/userlist.js
modified:   style.css
modified:   task-webix.html

LukshaMaria committed:
modified:   modules/maintable.js
changed to arrow function

LukshaMaria committed:
modified:   modules/forms.js
changed the method of adding value to the table

LukshaMaria committed:
Update forms.js

Deleted id:list.getLastId()+1

LukshaMaria committed:
modified:   modules/forms.js

LukshaMaria committed:
Merge branch 'webix-task-3' of https://github.com/LukshaMaria/Projects-Webix into webix-task-3

HelgaListopad commented on modules/userlist.js:
```diff
@@ -1,54 +1,55 @@
+                    function () {
+                        $$("userlist1").remove(id);
+                        webix.message("Deleted");
+                        return false;
```
Вот этот return лишний

LukshaMaria replied:
fixed

HelgaListopad commented on modules/maintable.js:
```diff
@@ -20,16 +20,14 @@ const maintable =
-                list.remove(id);
-                return false;
-       });
+            webix.confirm("Do you want to delete this item?", "confirm-warning").then(() => list.remove(id));
```
После webix.confirm не хватает return false (т.е. возвращать false нужно из обработчика клика wxi-trash)

LukshaMaria replied:
fixed

LukshaMaria committed:
Update userlist.js

Deleted return false.

LukshaMaria committed:
Update maintable.js

Added return false;
